### PR TITLE
drivers/sensors/sensor: fix poll()/read() for fetch()-only sensors

### PR DESCRIPTION
**Pull request by FelipeMdeO**
## Summary

A `fetch()`-only lower half reads the device on demand, so it's always
ready, nothing to wait for. The upper half didn't reflect that:
`sensor_poll()` only reported `POLLIN` when opened `O_NONBLOCK`, and a
blocking `read()` waited on `buffersem`, which only a driver-owned
interrupt (`notify_event`) posts. A `fetch()`-only sensor with no
interrupt therefore never satisfied `poll()`/`read()` at all — e.g.
in-tree `nucleo-h563zi:dts` hangs on a blocking `read()` today.
Applications worked around this with `O_NONBLOCK` (apache/nuttx-apps#3686);
this PR fixes it at the root instead, per reviewer feedback there.

Three commits:

1. **l3gd20 → push-only.** Its `BUFFER_SIZE == 0` mode set `fetch()` while
   still gating readiness on the data-ready interrupt. That mix was buggy,
   not just inconsistent: `poll()` reported `POLLIN` from `O_NONBLOCK`, not
   from the interrupt, so a multi-descriptor loop could read it stale or
   twice. Now always `push_event`, the path it already took by default; no
   in-tree config enables it.
2. **`sensor_poll()`/`sensor_read()` always ready for `fetch()`-only.** The
   actual fix. `buffersem` goes with it: its only two readers were the ones
   removed here, so the semaphore and its posts are now dead. `notify_event`
   is likewise dead for `fetch()`-only lower halves, but I left that API
   alone rather than widen the patch.
3. **Pace `POLLIN` at the requested interval.** A rate request got no help
   from `poll()`, and apps sleeping it out serialize across topics (three at
   10 Hz → 3.3 Hz each). A per-subscriber `wdog_s` armed in `sensor_poll()`
   paces each at its own interval. A watchdog rather than a work item keeps
   this off the work queue, which is what `fetch()` exists to avoid: it runs
   in timer context and needs no lock, so teardown just cancels it where it
   clears `fds`, and `sensor_close()` needs nothing of its own.

## Impact

Any `fetch()`-only uORB sensor (on-demand I2C/SPI, no interrupt) now
works with `poll()`/`read()` without the app forcing `O_NONBLOCK`, and
pacing costs no thread and no work queue — it needs no `CONFIG_` option
at all, where the earlier work-queue revision of this PR silently did
nothing without `CONFIG_SCHED_LPWORK`. Every interrupt-driven sensor,
including l3gd20 after commit 1, keeps blocking correctly until real new
data arrives; push semantics are unchanged, verified on both paths
(Testing).

## Testing

Host: Ubuntu 24.04.3 LTS. `checkpatch.sh` (style + `-m` commit messages)
clean on all 3 commits.

**Compiles clean** on 3 architectures touching the changed
`drivers/sensors/sensor.c`: xtensa-esp-elf-gcc 14.2.0 (`esp32s3-devkit`),
arm-none-eabi-gcc 13.2.1 (`stm32f4discovery`), and `sim` (x86_64).

### fetch()-only path — ESP32-S3-DevKitC + MPU6050 (GY-521) on I2C0

`fetch()`-only with no interrupt line, no `O_NONBLOCK` anywhere in the
unmodified `apps/system/uorb/listener.c`. Note this configuration does
**not** enable `CONFIG_SCHED_LPWORK`, so it is also the case the previous
work-queue revision could not have paced at all:

```
nsh> uorb_listener -n 5 -r 10 sensor_accel0
Monitor objects num:1
object_name:sensor_accel, object_instance:0
sensor_accel(now:25470000):timestamp:25470000,x:-0.160412,y:-0.191536,z:10.292673,temperature:24.106468
sensor_accel(now:25580000):timestamp:25580000,x:-0.134075,y:-0.093373,z:10.297462,temperature:24.012352
sensor_accel(now:25690000):timestamp:25690000,x:-0.138864,y:-0.134075,z:10.192117,temperature:23.918234
sensor_accel(now:25800000):timestamp:25800000,x:-0.105345,y:-0.155623,z:10.362105,temperature:24.012352
sensor_accel(now:25910000):timestamp:25910000,x:-0.124498,y:-0.148441,z:10.285490,temperature:24.012352
Object name:sensor_accel0, received:5
Total number of received Message:5/5
```

110 ms apart, the requested 10 Hz.

**Per-subscriber pacing**: two `uorb_listener` processes on the same
`sensor_accel0` at once, `-r 20` and `-r 5`. Their samples landed ~60 ms
and ~210 ms apart respectively — each paced at its own interval, not at
the minimum across both.

**Teardown under stress**: 10 rounds of `uorb_listener -r 5 sensor_accel0 &`
followed by `SIGINT` mid-poll. `ps` afterward shows no leaked task and no
armed timer, and `uorb_listener -n 3 sensor_accel0` right after still
completes 3/3.

### push path — sim, exercising the `buffersem` removal

`sim:baromonitor`, with `uorb_generator` as the publisher: `orb_publish()`
→ `write()` → `sensor_write()` → `sensor_push_event()`, which is exactly
where the `nxsem_post(&user->buffersem)` was removed.

```
nsh> uorb_generator -n 100 -r 5 -s -t sensor_baro0 timestamp:23191100,pressure:999.12,temperature:26.34 &
nsh> uorb_listener -n 5 sensor_baro0
Monitor objects num:1
object_name:sensor_baro, object_instance:0
sensor_baro(now:9070000):timestamp:8940000,pressure:999.119995,temperature:26.340000
sensor_baro(now:9150000):timestamp:9150000,pressure:999.119995,temperature:26.340000
sensor_baro(now:9360000):timestamp:9360000,pressure:999.119995,temperature:26.340000
sensor_baro(now:9570000):timestamp:9570000,pressure:999.119995,temperature:26.340000
sensor_baro(now:9780000):timestamp:9780000,pressure:999.119995,temperature:26.340000
Object name:sensor_baro0, received:5
Total number of received Message:5/5
```

The ~210 ms deltas track the publisher's 5 Hz, so the subscriber is being
woken by each push rather than spinning or stalling.

Two subscribers at once against a 10 Hz publisher, one of them rate
limited: the plain one got 6/6 ~110 ms apart, the `-r 2` one got 4/4
~500 ms apart — both notified from the same `sensor_push_event()` loop,
and the push-side rate gating in `sensor_is_updated()` is unaffected.

Six sequential subscribe/teardown cycles on that topic all completed and
exited cleanly, with no leaked task. That also covers an edge case of
commit 3: a push sensor never takes the branch that arms the watchdog, so
teardown's `wd_cancel()` always runs on a never-armed `wdog_s`. It is safe
by construction (`WDOG_ISACTIVE` tests `func != NULL`, and `user` comes
from `kmm_zalloc`), and this confirms it at runtime.

### CODE REVIEW

File: drivers/sensors/sensor.c [hunk 1]
```diff
@@ -1177,12 +1248,40 @@ static int sensor_poll(FAR struct file *filep,
...
+              uint64_t elapsed = now - user->fetched;
+
+              if (elapsed >= user->state.interval)
+                {
+                  user->fetched = now;
+                  eventset |= POLLIN;
+                }
+              else
+                {
+                  work_queue(LPWORK, &user->work, sensor_fetch_worker, user,
```
Comment (xiaoxiang781216) [inline drivers/sensors/sensor.c]: should we avoid the work queue? the benefit from fetch is avoiding the work or thread.

Comment (FelipeMdeO) [reply drivers/sensors/sensor.c]: Hello Xiang, I can't find a better solution.

Following my approach, the default use case is to use read or poll without an interval — in this case, no thread or work item. When the work item is used, it fires at the same cadence the caller wants to wake at anyway, so it's not extra cost, just moved from a userspace sleep to a kernel timer.
Another point is that orb_set_interval() already means "poll() gets ready about this often" for a pushing sensor. Without this, it silently means nothing for a fetch()-only one — a trap for anyone writing sensor-agnostic code against uORB.

File: drivers/sensors/sensor.c [hunk 2]
```diff
@@ -1184,20 +1177,12 @@ static int sensor_poll(FAR struct file *filep,
...
+           * anything to wait for.
+           */
 
-          if (filep->f_oflags & O_NONBLOCK)
-            {
-              eventset |= POLLIN;
-            }
-          else
-            {
-              nxsem_get_value(&user->buffersem, &semcount);
```
Comment (xiaoxiang781216) [inline drivers/sensors/sensor.c:1195]: should we remove buffersem too?

File: drivers/sensors/sensor.c [hunk 3]
```diff
@@ -96,6 +97,17 @@ struct sensor_user_s
...
   sensor_role_t    role;       /* The is used to indicate user's role based on open flags */
+#ifdef CONFIG_SCHED_LPWORK
+
+  /* Paces POLLIN for a fetch() only lower half at the interval this user
+   * asked for: upper is the device it belongs to, and fetched is when
+   * POLLIN was last reported to it, in microseconds.
+   */
+
+  FAR struct sensor_upperhalf_s *upper;
+  struct work_s    work;
```
Comment (xiaoxiang781216) [inline drivers/sensors/sensor.c]: let's change to wdog

File: drivers/sensors/sensor.c [hunk 4]
```diff
@@ -809,6 +867,19 @@ static int sensor_close(FAR struct file *filep)
...
 
+#ifdef CONFIG_SCHED_LPWORK
+  /* A close() racing an armed poll() would otherwise leave the worker
+   * queued against the user we are about to free. Clear fds first so a
+   * worker already running cannot re-arm itself past the cancel, then
+   * cancel outside the lock, since the worker takes it too.
+   */
+
+  nxrmutex_lock(&upper->lock);
+  user->fds = NULL;
```
Comment (xiaoxiang781216) [inline drivers/sensors/sensor.c]: why need? before close, sensor_poll with false should be called first.

File: drivers/sensors/sensor.c [hunk 5]
```diff
@@ -1204,6 +1303,19 @@ static int sensor_poll(FAR struct file *filep,
...
+
+#ifdef CONFIG_SCHED_LPWORK
+  /* Cancel outside the lock: the worker takes it, so cancelling while
+   * holding it would deadlock. fds was cleared above, so a worker that is
+   * already running will neither notify nor re-arm.
+   */
+
+  if (!setup)
+    {
+      work_cancel_sync(LPWORK, &user->work);
```
Comment (xiaoxiang781216) [inline drivers/sensors/sensor.c]: move after line 1301

File: drivers/sensors/sensor.c (hunk 1 above)
Comment (xiaoxiang781216) [reply drivers/sensors/sensor.c]: OK, let's continue.

File: drivers/sensors/sensor.c (hunk 5 above)
Comment (FelipeMdeO) [reply drivers/sensors/sensor.c]: Ok, I will

File: drivers/sensors/sensor.c (hunk 2 above)
Comment (FelipeMdeO) [reply drivers/sensors/sensor.c:1195]: Yes, we should remove buffersem also, I will remove.

File: drivers/sensors/sensor.c (hunk 3 above)
Comment (FelipeMdeO) [reply drivers/sensors/sensor.c]: wdog is the right approach, thanks for highlight it.

File: drivers/sensors/sensor.c (hunk 4 above)
Comment (FelipeMdeO) [reply drivers/sensors/sensor.c]: Removed after implementation using wdog

**Comment by FelipeMdeO** [comment]
Hello Xiang, I updated implementation to use wdog than LPWORKER. Also I updated the description to follow the new approach. Thanks for suggest wdog.

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by FelipeMdeO** [comment]
Hello @linguini1 , can you take a look, please? I think you used uOrb recently, right?

**Review by jerpelea — APPROVED** [review]
